datawig/hpo.py

```diff
@@ -0,0 +1,400 @@
+            self.hps['concat'] = {}
+
+        # add type to column dictionaries if it was not specified, does not support categorical types
+        for column_name in simple_imputer.input_columns:
+            if column_name not in self.hps.keys():
+                self.hps[column_name] = {}
+            if 'type' not in self.hps[column_name].keys():
+                if is_numeric_dtype(train_df[column_name]):
+                    self.hps[column_name]['type'] = ['numeric']
+                else:
+                    self.hps[column_name]['type'] = ['string']
+
+        # check that all passed parameter are valid hyperparameters
+        assert all([key in self.default_hps['global'].keys() for key in self.hps['global'].keys()])
+        for key, val in self.hps.items():
+            if 'type' in val.keys():
+                assert all([key in list(self.default_hps[val['type'][0]].keys()) + ['type'] for key, _ in val.items()])
+
+        # augment provided global self.hps with default global hps so that cartesian products are full parameter sets.
+        # self.hps['global'] = {**self.default_hps['global'], **self.hps['global']}
+        self.hps['global'] = merge_dicts(self.default_hps['global'], self.hps['global'])
+        self.hps['concat'] = merge_dicts(self.default_hps['concat'], self.hps['concat'])
+
+        # augment provided self.hps with default self.hps, iterating over every input column
+        for column_name in simple_imputer.input_columns:
+            # add dictionary for input columns where no hyperparamters have been passed
+            if column_name not in self.hps.keys():
+                self.hps[column_name] = {}
+                if column_name in simple_imputer.numeric_columns:
+                    self.hps[column_name]['type'] = ['numeric']
+                elif column_name in simple_imputer.string_columns:
+                    self.hps[column_name]['type'] = ['string']
+                else:
+                    logger.warn('Input type of column ' + str(column_name) + ' not determined.')
+            # join all column specific hp dictionaries with type-specific default values
+            # self.hps[column_name] = {**self.default_hps[self.hps[column_name]['type'][0]], **self.hps[column_name]}
+            self.hps[column_name] = merge_dicts(self.default_hps[self.hps[column_name]['type'][0]],
+                                                self.hps[column_name])
+
+        # flatten nested dictionary structures and combine to data frame with all possible hp configurations
+        hp_df_from_dict = lambda dict: pd.DataFrame(list(itertools.product(*dict.values())), columns=dict.keys())
+
+        return hp_df_from_dict(flatten_dict(self.hps))
+
+    def __fit_hp(self,
+                 train_df: pd.DataFrame,
+                 test_df: pd.DataFrame,
+                 hp: pd.Series,
+                 simple_imputer,
+                 user_defined_scores: list = None,
+                 name: str = ''):
```

Comment:
if name is an empty string we're overwriting the results of each of the hpo runs? maybe it's better to remove the default value and check that it is not empty.